aspects/schema.go

```diff
@@ -0,0 +1,358 @@
+	return s.topLevel.Validate(raw)
+}
+
+func (s *StorageSchema) parse(raw json.RawMessage) (parser, error) {
+	var typ string
+	var schemaDef map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &schemaDef); err != nil {
+		var typeErr *json.UnmarshalTypeError
+		if !errors.As(err, &typeErr) {
+			return nil, err
+		}
+
+		if err := json.Unmarshal(raw, &typ); err != nil {
+			return nil, err
+		}
+	} else {
+		rawType, ok := schemaDef["type"]
+		if !ok {
+			typ = "map"
+		} else {
+			if err := json.Unmarshal(rawType, &typ); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	schema, err := s.newTypeSchema(typ)
+	if err != nil {
+		return nil, err
+	}
+
+	// only parse the schema if it's a map definition w/ constraints
+	if schemaDef != nil {
+		if err := schema.parseConstraints(schemaDef); err != nil {
+			return nil, err
+		}
+	}
+
+	return schema, nil
+}
+
+func (s *StorageSchema) newTypeSchema(typ string) (parser, error) {
+	// TODO: add any, int, number, bool, array and user-defined types
+	switch typ {
+	case "map":
+		return &mapSchema{topSchema: s}, nil
+	case "string":
+		return &stringSchema{}, nil
+	default:
+		return nil, fmt.Errorf("cannot parse type %q: unknown", typ)
+	}
+}
+
+type mapSchema struct {
+	// topSchema is the schema for the top-level schema which contains the user types.
+	topSchema *StorageSchema
+
+	// entrySchemas maps keys to their expected types. Alternatively, the schema
+	// can constrain key and/or value types.
+	entrySchemas map[string]Schema
+
+	// valueSchema validates that the map's values match a certain type.
+	valueSchema Schema
+
+	// keySchema validates that the map's key match a certain type.
+	keySchema Schema
+
+	// requiredCombs holds combinations of keys that an instance of the map is
+	// allowed to have.
+	requiredCombs [][]string
+}
+
+// Validate that raw is a valid aspect map and meets the constraints set by the
+// aspect schema.
+func (v *mapSchema) Validate(raw []byte) error {
+	var mapValue map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &mapValue); err != nil {
+		return err
+	}
+
+	var missing bool
+	for _, required := range v.requiredCombs {
+		missing = false
+		for _, key := range required {
+			if _, ok := mapValue[key]; !ok {
+				missing = true
+				break
+			}
+		}
+
+		if !missing {
+			// matched possible combination of required keys so we can stop
+			break
+		}
+	}
+
+	if missing {
+		return fmt.Errorf(`cannot find required combinations of keys`)
+	}
+
+	if v.entrySchemas != nil {
+		for key, val := range mapValue {
+			if validator, ok := v.entrySchemas[key]; ok {
+				if err := validator.Validate(val); err != nil {
+					return err
+				}
+			}
+		}
+
+		// all required entries are present and validated
+		return nil
+	}
+
+	if v.keySchema != nil {
+		for k := range mapValue {
+			rawKey, err := json.Marshal(k)
+			if err != nil {
+				return err
+			}
+
+			if err := v.keySchema.Validate(rawKey); err != nil {
+				return err
```

Comment:
do we need a test that hits this error path? or is not possible atm because we don't have string contraints? should we leave a todo?